File: devicesetup/src/main/java/io/particle/android/sdk/utils/SEGAnalytics.java
package io.particle.android.sdk.utils;

import android.content.Context;

import com.segment.analytics.Analytics;
import com.segment.analytics.Properties;


/**
 * Created by Julius.
 */
public class SEGAnalytics {
    public static String analyticsKey = "";
    public static boolean analyticsOptOut = true;

    public static void initialize(Context context) {
        //FIXME need to find method call whether analytics are set or export this part to application class 
        try {
            Analytics.with(context);
        } catch (IllegalArgumentException exception) {
            // Create an analytics client with the given context and Segment write key. 
            if (!analyticsKey.isEmpty()) {
                Analytics analytics = new Analytics.Builder(context, analyticsKey).build();
                analytics.optOut(analyticsOptOut);
                Analytics.setSingletonInstance(analytics);
            }
        }
    }

    public static void track(Context context, String track) {
        if (!analyticsKey.isEmpty()) {
            Analytics.with(context).track(track);
        }
    }

    public static void screen(Context context, String screen) {
        if (!analyticsKey.isEmpty()) {
            Analytics.with(context).track(screen);
        }
    }

    public static void track(Context context, String track, Properties analyticProperties) {
        if (!analyticsKey.isEmpty()) {
            Analytics.with(context).track(track, analyticProperties);
        }
    }

    public static void identify(Context context, String email) {
        if (!analyticsKey.isEmpty()) {
            Analytics.with(context).identify(email);
        }
    }
}

File: devicesetup/src/main/java/io/particle/android/sdk/utils/SEGAnalytics.kt
package io.particle.android.sdk.utils

import android.annotation.SuppressLint
import android.content.Context
import android.support.annotation.RestrictTo

import com.segment.analytics.Analytics
import com.segment.analytics.Properties


/**
 * Created by Julius.
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
object SEGAnalytics {
    var analyticsKey = ""
    var analyticsOptOut = true
    lateinit var analytics: Analytics
    @SuppressLint("StaticFieldLeak")

    fun initialize(context: Context) {
        try {
            analytics = Analytics.with(context)
        } catch (exception: IllegalArgumentException) {
            if (!analyticsKey.isEmpty()) {
                val analytics = Analytics.Builder(context, analyticsKey).build()
                analytics.optOut(analyticsOptOut)
                Analytics.setSingletonInstance(analytics)
            }
        }

    }

    fun track(track: String) {
        if (!analyticsKey.isEmpty()) {
            analytics.track(track)
        }
    }

    fun screen(screen: String) {
        if (!analyticsKey.isEmpty()) {
            analytics.track(screen)
        }
    }

    fun track(track: String, analyticProperties: Properties?) {
        if (!analyticsKey.isEmpty()) {
            analytics.track(track, analyticProperties)
        }
    }

    fun identify(email: String?) {
        if (!analyticsKey.isEmpty() && !email.isNullOrEmpty()) {
            analytics.identify(email!!)
        }
    }
}
